components/supervisor/pkg/supervisor/tasks.go

```diff
@@ -0,0 +1,234 @@
+	config            *Config
+	tasks             map[string]*api.TasksStatus
+	ready             chan struct{}
+	terminalService   *terminal.MuxTerminalService
+	inWorkspaceHelper *backup.InWorkspaceHelper
+}
+
+func newTasksManager(config *Config, terminalService *terminal.MuxTerminalService, inWorkspaceHelper *backup.InWorkspaceHelper) *tasksManager {
+	return &tasksManager{
+		config:            config,
+		terminalService:   terminalService,
+		inWorkspaceHelper: inWorkspaceHelper,
+		tasks:             make(map[string]*api.TasksStatus),
+		ready:             make(chan struct{}),
+	}
+}
+
+func (tm *tasksManager) Run(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+	defer close(tm.ready)
+
+	tasks, err := tm.config.getGitpodTasks()
+	if err != nil {
+		log.WithError(err).Fatal()
+		return
+	}
+	if tasks == nil {
+		log.Info("no gitpod tasks found")
+		return
+	}
+
+	select {
+	case <-ctx.Done():
+		return
+	case <-tm.inWorkspaceHelper.ContentReady():
+		contentSource, _ := tm.inWorkspaceHelper.ContentSource()
+		headless := tm.config.GitpodHeadless != nil && *tm.config.GitpodHeadless == "true"
+		for i, task := range *tasks {
+			index := strconv.Itoa(i)
+			prebuildLogFileName := "/workspace/.gitpod/prebuild-log-" + index
+
+			var taskCommands []*string
+			if headless {
+				// prebuild
+				taskCommands = []*string{task.Before, task.Init, task.Prebuild}
+			} else if contentSource == csapi.WorkspaceInitFromPrebuild {
+				// prebuilt
+				legacyPrebuildLogFileName := "/workspace/.prebuild-log-" + index
+				printlogs := "[ -r " + legacyPrebuildLogFileName + " ] && cat " + legacyPrebuildLogFileName + "; [ -r " + prebuildLogFileName + " ] && cat " + prebuildLogFileName + "; true"
+				taskCommands = []*string{task.Before, &printlogs, task.Command}
+			} else if contentSource == csapi.WorkspaceInitFromBackup {
+				// restart
+				taskCommands = []*string{task.Before, task.Command}
+			} else {
+				// init
+				taskCommands = []*string{task.Before, task.Init, task.Command}
+			}
+			if taskCommands == nil {
+				continue
+			}
+			command := reduce(taskCommands, func(composed string, command string) string {
+				if composed != "" {
+					composed += " && "
+				}
+				return composed + "{\n" + command + "\n}"
+			})
+
+			if headless {
+				// it's important that prebuild tasks exit eventually
+				// also, we need to save the log output in the workspace
+				if strings.TrimSpace(command) == "" {
+					command = "exit"
+				} else {
+					command += "; exit"
+				}
+			} else if strings.TrimSpace(command) != "" {
+				histfile := "/workspace/.gitpod/cmd-" + index
+				histfileCommands := taskCommands
+				if contentSource == csapi.WorkspaceInitFromPrebuild {
+					histfileCommands = []*string{task.Before, task.Init, task.Prebuild, task.Command}
+				}
+				err = ioutil.WriteFile(histfile, []byte(reduce(histfileCommands, func(composed string, command string) string {
+					return composed + command + "\r\n"
+				})), 0644)
```

Comment:
here, you could use the `cmds` from above - again removing the need for `reduce`